xbmc/filesystem/BlurayDirectory.cpp

```diff
@@ -66,49 +66,540 @@ std::string CBlurayDirectory::GetDiscInfoString(DiscInfo info)
+#if (BLURAY_VERSION > BLURAY_VERSION_CODE(1, 0, 0))
+      title = disc_info->disc_name ? disc_info->disc_name : "";
 #endif
 
-    return title;
+      return title;
+    }
+    case XFILE::CBlurayDirectory::DiscInfo::ID:
+    {
+      if (!m_blurayInitialized)
+        return "";
+
+      const BLURAY_DISC_INFO* disc_info = bd_get_disc_info(m_bd);
+      if (!disc_info || !disc_info->bluray_detected)
+        return "";
+
+      std::string id = "";
+
+#if (BLURAY_VERSION > BLURAY_VERSION_CODE(1, 0, 0))
+      id = disc_info->udf_volume_id ? disc_info->udf_volume_id : "";
+
+      if (id.empty())
+      {
+        id = HexToString(disc_info->disc_id, 20);
+      }
+#endif
+
+      return id;
+    }
+    default:
+      break;
   }
-  case XFILE::CBlurayDirectory::DiscInfo::ID:
+
+  return "";
+}
+
+static constexpr unsigned int MIN_EPISODE_LENGTH = 15 * 60; // 15 minutes
+static constexpr unsigned int MAX_EPISODE_DIFFERENCE = 30; // 30 seconds
+
+void CBlurayDirectory::GetEpisodeTitles(const CFileItem& episode,
+                                        CFileItemList& items,
+                                        const std::vector<CVideoInfoTag>& episodesOnDisc)
+{
+  // Get all titles on disc
+  // Sort by playlist for grouping later
+  CFileItemList allTitles;
+  GetTitles(false, allTitles);
+  SortDescription sorting;
+  sorting.sortBy = SortByFile;
+  allTitles.Sort(sorting);
+
+  // Find our episode on disc
+  unsigned int episodeOffset = 0;
+  for (int i = 0; i < (int)episodesOnDisc.size(); ++i)
   {
-    if (!m_blurayInitialized)
-      return "";
+    if (episodesOnDisc[i].m_iEpisode == episode.GetVideoInfoTag()->m_iEpisode)
+    {
+      episodeOffset = i;
+      break;
+    }
+  }
+
+  const unsigned int numEpisodes = (int)episodesOnDisc.size();
+  const unsigned int episodeDuration = episode.GetVideoInfoTag()->GetDuration();
+
+  CLog::Log(LOGDEBUG, "*** Episode Search Start ***");
+
+  CLog::Log(LOGDEBUG, "Looking for episode {} duration {}", episode.GetVideoInfoTag()->m_iEpisode,
+            episode.GetVideoInfoTag()->GetDuration());
 
-    const BLURAY_DISC_INFO* disc_info = bd_get_disc_info(m_bd);
-    if (!disc_info || !disc_info->bluray_detected)
-      return "";
+  // List episodes expected on disc
+  for (const auto& e : episodesOnDisc)
+  {
+    CLog::Log(LOGDEBUG, "On disc - episode {} duration {}", e.m_iEpisode, e.GetDuration());
+  }
 
-    std::string id = "";
+  // Get information on all titles
+  // Including relationship between clips and playlists
+  // List all playlists
+  std::vector<std::vector<unsigned int>> clips;
+  std::vector<std::vector<unsigned int>> playlists;
+  std::vector<std::pair<unsigned int, std::string>> playlist_langs;
 
-#if (BLURAY_VERSION > BLURAY_VERSION_CODE(1,0,0))
-    id = disc_info->udf_volume_id ? disc_info->udf_volume_id : "";
+  for (const auto& title : allTitles)
+  {
+    CFileItemPtr newItem(new CFileItem("", false));
```

Comment:
`const auto newItem{std::make_shared<CFileItem>("", false)};`

Reply:
Line not atcually needed. Removed.